### PR TITLE
Fix perf when streams don't change often

### PR DESCRIPTION
**Pull request by erikjohnston**
There is a bug with the `StreamChangeCache` where it would incorrectly return that all entities had changed if asked for entities changed *since* the earliest stream position.

Note that for streams we use the inequalities: `$min_stream_id < stream_id <= $max_stream_id`, i.e. when we ask the stream change cache for all things that have changed since `$stream_id` we don't care for events that happened *at* `$stream_id`.

Specifically: `_earliest_known_stream_pos` is the position at which we know that we'll have entries for all changes since that point, we can use the cache for any stream IDs that equal `_earliest_known_stream_pos`.

`_earliest_known_stream_pos` is set in three places:
- On startup we set it either to:
  - the current maximum stream ID, with not prefilled values; or
  - the minimum of the latest N values we pulled from the DB
- When we evict items from the bottom, we set it to the stream ID of the evicted items.

This was changed in https://github.com/matrix-org/synapse/pull/14435, but I think we were overly conservative there.

### CODE REVIEW

File: synapse/util/caches/stream_change_cache.py [hunk 1]
```diff
@@ -142,9 +142,9 @@ def has_entity_changed(self, entity: EntityType, stream_pos: int) -> bool:
         """
         assert isinstance(stream_pos, int)
 
-        # _cache is not valid at or before the earliest known stream position, so
+        # _cache is not valid before the earliest known stream position, so
         # return that the entity has changed.
-        if stream_pos <= self._earliest_known_stream_pos:
+        if stream_pos < self._earliest_known_stream_pos:
```
Comment (clokep) [inline synapse/util/caches/stream_change_cache.py:147]: We had a long conversation about this @ https://github.com/matrix-org/synapse/pull/14435#discussion_r1021860786, during which I said:

> I'm a bit skeptical of the resulting consistency between strict and non-strict comparisons, but would like a double checking.

Your response was:

> as if we call `entity_has_changed` for a stream pos then `has_entity_changed` *must* return True for that stream pos. Either because we remember the entity has changed or because the stream pos is before the minimum threshold

It looks like I interpreted "before" incorrectly in that case, since *at* the stream position we do have the "proper" information needed to calculate whether there were changes or not.

The comment for `_earliest_known_stream_pos` even makes this clear:

https://github.com/element-hq/synapse/blob/ef9ef99f592a62949ce49fca0a912e9b8892aed6/synapse/util/caches/stream_change_cache.py#L91-L93

tl;dr Before #14435 we were inconsistently checking if it was strictly before or before-or-equal, we made them consistent in that PR, but to the wrong one.

File: tests/util/test_stream_change_cache.py [hunk 2]
```diff
@@ -53,8 +53,8 @@ def test_has_entity_changed(self) -> None:
         # return True, whether it's a known entity or not.
         self.assertTrue(cache.has_entity_changed("user@foo.com", 0))
         self.assertTrue(cache.has_entity_changed("not@here.website", 0))
-        self.assertTrue(cache.has_entity_changed("user@foo.com", 3))
-        self.assertTrue(cache.has_entity_changed("not@here.website", 3))
+        self.assertTrue(cache.has_entity_changed("user@foo.com", 2))
+        self.assertTrue(cache.has_entity_changed("not@here.website", 2))
```
Comment (clokep) [inline tests/util/test_stream_change_cache.py:57]: Should we add checks for `3` (near line 35)?

**Review by clokep — APPROVED** [review]
I believe this looks correct, I think it is missing 1 tweak to the tests though.

File: changelog.d/17767.misc [hunk 3]
```diff
@@ -0,0 +1 @@
+Fix perf when streams don't change often.
```
Comment (anoadragon453) [inline changelog.d/17767.misc:1]:
```suggestion
Fix performance of streams that don't change often.
```
